File: update_lexicon.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Lookup each entry in lexicon-dict mapping and replace
# declarations in lexicon with corresponding ones from dict

import sys
import codecs
import re

MAP="lex-dict-mapping.txt"
LEX="LexiconMlt.gf"
DIC="DictMlt.gf"

if __name__ == "__main__":

    # Load mapping into a dictionary
    mapfile = codecs.open(MAP,'r','utf-8')
    d_map = {}
    for l in mapfile.readlines():
        m = l.split("=")
        k = m[0].strip()
        v = m[1].strip()
        d_map[k] = v
    mapfile.close();

    # Load Dict into a dictionary
    dicfile = codecs.open(DIC,'r','utf-8')
    d_dict = {}
    for l in dicfile.readlines():
        m = l.split("=")
        if (len(m)) < 2:
            continue
        k = m[0].strip()
        v = m[1].strip()
        d_dict[k] = v
    dicfile.close();

    # Replace declarations with those in Dict file
    for (k,v) in d_map.items():
        if v in d_dict:
            d_map[k] = d_dict[v]

    # Replace declarations in Lexicon file
    lexfile = codecs.open(LEX,'r','utf-8')
    c = lexfile.read()
    for (k,v) in d_map.items():
        c = re.sub(k+r'.+;', k+r' = '+v, c)
    lexfile = codecs.open(LEX,'w','utf-8')
    lexfile.write(c)
    lexfile.close()

File: update_lexicon.hs
-- Lookup each entry in lexicon-dict mapping and replace
-- declarations in lexicon with corresponding ones from dict

import Text.Regex
import Data.Maybe

mapFile="lex-dict-mapping.txt"
dicFile="DictMlt.gf"
lexFile="LexiconMlt.gf"

-- Strip whitespace from a string
strip = lstrip . rstrip
lstrip = dropWhile (`elem` " \t")
rstrip = reverse . lstrip . reverse

-- Read file into lines, splitting each at given separator
readAndSplitAt :: FilePath -> String -> IO [(String,String)]
readAndSplitAt file sep = do
  content <- readFile file
  return [ (strip (x!!0), strip (x!!1))
         | x <- map (splitRegex (mkRegex sep)) (lines content)
         , length x > 1]

main = do
  -- Load map and construct replacements
  ms <- readAndSplitAt mapFile "="
  ds <- readAndSplitAt dicFile "="
  let rs = [ (k, fromJust (lookup v1 ds))
          | (k,v1) <- ms
          , isJust (lookup v1 ds) ]

  -- Perform replacements
  content <- readFile lexFile
  let ls' = map replaceLine (lines content)
      replaceLine :: String -> String
      replaceLine l =
        let k = strip $ takeWhile (/='=') l
        in case lookup k rs of
          Just v -> "    " ++ k ++ " = " ++ v -- ++ " -- auto-replaced"
          _ -> l
  
  -- Output contents
  -- writeFile lexFile (unlines ls')
  putStr $ unlines ls'
  
